pkg/storage/postgres/client.go

```diff
@@ -53,6 +53,7 @@ func (c *Client) Close() error {
 }
 
 type dbConn interface {
+	BeginFunc(context.Context, func(pgx.Tx) error) error
```

Comment:
This looks a bit inconsistent to me. We can use this interface to create transaction, but the other methods execute outside of transactions. I would like to propose an approach for this:

```go
type txMan interface {
  BeginTx() tx
}

type tx interface {
  QueryRow(context.Context, string, ...interface{}) pgx.Row
  Query(context.Context, string, ...interface{}) (pgx.Rows, error)
  Exec(context.Context, string, ...interface{}) (pgconn.CommandTag, error)
  Commit() error
  Rollback() error
}
```

If adopted, the `storage` package would be changed to instead of having a `Storer` interface, having instead a `TxManager` and a `Tx` interface. All requests would go through a transaction, and all methods would be changed to receive a transaction instead of a `Storer`. This would have the (imo) huge advantage of not leaving garbage behind when storing runs and other objects that use multiple records on multiple tables.

Reply:
The [storer interface](https://github.com/simplesurance/baur/blob/1408113e5cb3f698ec024868f4cf679f2b363c7a/pkg/storage/storage.go#L93) does not have to be aware of transactions. Transactions are imho an implementation detail of this level. The methods of this interface should ensure that data is consistent, however this is done internally.
All operations that store data (it's actually only one https://github.com/simplesurance/baur/blob/1408113e5cb3f698ec024868f4cf679f2b363c7a/pkg/storage/postgres/insert.go#L510-L517) are already using transactions.

Does this make sense or do you have something else in mind?

Reply:
The use case to which I was proposing the use of a transaction manager don't actually match baur very well, since the storage interface have "big methods", that receive complete objects and store them atomically. The transaction manager would fit the case where there are smaller methods, and where multiple calls have to be done consistently, withing the same transaction. Because of that I will withdraw my suggestion.